README.md
# StackTracey

[![Build Status](https://travis-ci.org/xpl/stacktracey.svg?branch=master)](https://travis-ci.org/xpl/stacktracey) [![Windows Build Status](https://ci.appveyor.com/api/projects/status/cvuygb8grrvm1sdm?svg=true)](https://ci.appveyor.com/project/xpl/stacktracey) [![Coverage Status](https://coveralls.io/repos/github/xpl/stacktracey/badge.svg?branch=master)](https://coveralls.io/github/xpl/stacktracey) [![NPM](https://img.shields.io/npm/v/stacktracey.svg)](http://npmjs.com/package/stacktracey) [![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/xpl/stacktracey.svg)](https://scrutinizer-ci.com/g/xpl/stacktracey/?branch=master) [![dependencies Status](https://david-dm.org/xpl/stacktracey/status.svg)](https://david-dm.org/xpl/stacktracey)

Parses call stacks. Reads sources. Clean & filtered output. Sourcemaps. Node & browsers.

## Why

- [x] Simple
- [x] Works in Node and browsers, \*nix and Windows
- [x] Allows hiding library calls / ad-hoc exclusion (via `// @hide` marker)
- [x] Provides source text for call locations
- [x] Fetches sources (via [get-source](https://github.com/xpl/get-source))
- [x] Supports both asynchronous and synchronous interfaces (works even in browsers)
- [x] Full sourcemap support
- [x] Extracts useful information from `SyntaxError` instances
- [x] [Pretty printing](https://github.com/xpl/stacktracey/#pretty-printing)
      <img width="898" alt="screen shot 2017-09-27 at 16 53 46" src="https://user-images.githubusercontent.com/1707/30917345-79899004-a3a4-11e7-8d48-e217e2d5e2cd.png">

## What For

- [Error overlay UIs](https://github.com/xpl/panic-overlay/#how-it-looks) for easier front-end development
- [Better error reporting](https://github.com/xpl/ololog#pretty-printing-error-instances) for Node projects
- [Advanced logging](https://github.com/xpl/ololog#displaying-call-location) (displaying call locations)
- Assertion printing

## How To

```bash
npm install stacktracey
```

```javascript
import StackTracey from 'stacktracey'
```

Captures the current call stack:

```javascript
stack = new StackTracey ()            // captures the current call stack
```

Parses stacks from an `Error` object:

```javascript
stack = new StackTracey (error)
stack = new StackTracey (error.stack) // ...or from raw string
```

Stores parsed data in `.items`:

```javascript
stack.items.length // num entries
stack.items[0]     // top
```

...where each item exposes:

```javascript
{
    beforeParse:  <original text>,
    callee:       <function name>,
    calleeShort:  <shortened function name>,
    file:         <full path to file>,       // e.g. /Users/john/my_project/node_modules/foobar/main.js
    fileRelative: <relative path to file>,   // e.g. node_modules/foobar/main.js
    fileShort:    <short path to file>,      // e.g. foobar/main.js
    fileName:     <file name>,               // e.g. main.js
    line:         <line number>,             // starts from 1
    column:       <column number>,           // starts from 1

    index:          /* true if occured in HTML file at index page    */,
    native:         /* true if occured in native browser code        */,
    thirdParty:     /* true if occured in library code               */,
    hide:           /* true if marked as hidden by "// @hide" tag    */,
    syntaxError:    /* true if generated from a SyntaxError instance */
}
```

Accessing sources (**synchronously**, use with caution in browsers):

```javascript
stack = stack.withSources () // returns a copy of stack with all items supplied with sources
top   = stack.items[0]       // top item
```

Accessing sources (**asynchronously**, preferred method in browsers):

```javascript
stack = await stack.withSourcesAsync () // returns a copy of stack with all items supplied with sources
top   = stack.items[0]                  // top item
```

...or:

```javascript
top = stack.withSourceAt (0) // supplies source for an individiual item (by index)
```
```javascript
top = await stack.withSourceAsyncAt (0) // supplies source for an individiual item (by index)
```

...or:

```javascript
top = stack.withSource (stack.items[0]) // supplies source for an individiual item
```
```javascript
top = await stack.withSourceAsync (stack.items[0]) // supplies source for an individiual item
```

The returned items contain the following additional fields (already mapped through sourcemaps):

```javascript
{
    ... // all the previously described fields

    line:       <original line number>,
    column:     <original column number>,
    sourceFile: <original source file object>,
    sourceLine: <original source line text>
}
```

To learn about the `sourceFile` object, read the [get-source](https://github.com/xpl/get-source#get-source) docs.

## Cleaning Output

Synchronously (use with caution in browsers):

```javascript
stack = stack.clean ()
```

...or (asynchronously):

```javascript
stack = await stack.cleanAsync ()
```

It does the following:

1. Reads sources (if available)
2. Excludes locations marked with the `isThirdParty` flag (library calls)
3. Excludes locations marked with a `// @hide` comment (user defined exclusion)
4. Merges repeated lines (via the `.mergeRepeatedLines`)

You can customize its behavior by overriding the `isClean (entry, index)` predicate.

## Custom `isThirdParty` Predicate

You can override the `isThirdParty` behavior by subclassing `StackTracey`:

```javascript
class MyStackTracey extends StackTracey {

    isThirdParty (path) {
        return (super.isThirdParty (path)    // include default behavior
                || path.includes ('my-lib')) // paths including 'my-lib' will be marked as thirdParty
                && !path.includes ('jquery') // jquery paths won't be marked as thirdParty
    }
}

...

const stack = new MyStackTracey (error).withSources ()
```

## Pretty Printing

```javascript
const prettyPrintedString = new StackTracey (error).withSources ().asTable ()
```
```javascript
const prettyPrintedString = (await new StackTracey (error).withSourcesAsync ()).asTable () // asynchronous version
```

...or (for pretty printing cleaned output):

```javascript
const prettyPrintedString = new StackTracey (error).clean ().asTable ()
```
```javascript
const prettyPrintedString = (await new StackTracey (error).cleanAsync ()).asTable () // asynchronous version
```

It produces a nice compact table layout (thanks to [`as-table`](https://github.com/xpl/as-table)), supplied with source lines (if available):

```
at shouldBeVisibleInStackTrace     test.js:25                 const shouldBeVisibleInStackTrace = () => new StackTracey ()
at it                              test.js:100                const stack = shouldBeVisibleInStackTrace ()                
at callFn                          mocha/lib/runnable.js:326  var result = fn.call(ctx);                                  
at run                             mocha/lib/runnable.js:319  callFn(this.fn);                                            
at runTest                         mocha/lib/runner.js:422    test.run(fn);                                               
at                                 mocha/lib/runner.js:528    self.runTest(function(err) {                                
at next                            mocha/lib/runner.js:342    return fn();                                                
at                                 mocha/lib/runner.js:352    next(suites.pop());                                         
at next                            mocha/lib/runner.js:284    return fn();                                                
at <anonymous>                     mocha/lib/runner.js:320    next(0);                  
```

If you find your pretty printed tables undesirably trimmed (or maybe too long to fit in the line), you can provide custom column widths when calling `asTable` (...or, alternatively, by overriding `maxColumnWidths ()` method):

```javascript
stack.asTable ({
    callee:     30,
    file:       60,
    sourceLine: 80
})
```

## Using As A Custom Exception Printer In Node

You can even replace the default NodeJS exception printer with this! This is how you can do it:

```javascript
process.on ('uncaughtException',  e => { /* print the stack here */ })
process.on ('unhandledRejection', e => { /* print the stack here */ })
```

But the most simple way to achieve that is to use the [`ololog`](https://github.com/xpl/ololog/blob/master/README.md) library (that is built upon StackTracey and several other handy libraries coded by me). Check it out, [it's pretty awesome and will blow your brains out](https://github.com/xpl/ololog/blob/master/README.md) :)

```javascript
const log = require ('ololog').handleNodeErrors ()

// you can also print Errors by simply passing them to the log() function
```

<img width="1066" alt="screen shot 2018-05-11 at 19 51 03" src="https://user-images.githubusercontent.com/1707/39936393-ffd529c2-5554-11e8-80f8-eff1229017c4.png">
 
## Parsing `SyntaxError` instances

For example, when trying to `require` a file named `test_files/syntax_error.js`:

```javascript
// next line contains a syntax error (not a valid JavaScript)
foo->bar ()
```

...the pretty printed call stack for the error thrown would be something like:

```
at (syntax error)                  test_files/syntax_error.js:2  foo->bar ()
at it                              test.js:184                   try { require ('./test_files/syntax_error.js') }
at runCallback                     timers.js:781
at tryOnImmediate                  timers.js:743
at processImmediate [as _immediat  timers.js:714
```

...where the first line is generated from parsing the raw output from the `util.inspect` call in Node. Unfortunately, this won't work in older versions of Node (v4 and below) as these versions can't provide any meaningful information for a `SyntaxError` instance.

## Array Methods

All StackTracey instances expose `map`, `filter`, `concat` and `slice` methods. These methods will return mapped, filtered, joined, reversed and sliced `StackTracey` instances, respectively:

```javascript
s = new StackTracey ().slice (1).filter (x => !x.thirdParty) // current stack shifted by 1 and cleaned from library calls

s instanceof StackTracey // true
```

## Extra Stuff

You can compare two locations via this predicate (tests `file`, `line` and `column` for equality):

```javascript
StackTracey.locationsEqual (a, b)
```

To force-reload the sources, you can invalidate the global source cache:

```javascript
StackTracey.resetCache ()
```

## Projects That Use StackTracey

- [Ololog](https://github.com/xpl/ololog) — a better `console.log` for the log-driven debugging junkies!
- [CCXT](https://github.com/ccxt-dev/ccxt) — a cryptocurrency trading library that supports 130+ exchanges
- [pnpm](https://github.com/pnpm/pnpm) — a fast, disk space efficient package manager (faster than npm and Yarn!)
- [panic-overlay](https://github.com/xpl/panic-overlay/) — a lightweight standalone alternative to `react-error-overlay`



CHANGELOG.md
# Version 2.0 (with breaking changes)

- Added asynchronous methods, `withSourcesAsync` and `cleanAsync`.

- All property accessors like `.clean` are now methods (e.g. `.clean ()`) for consistency reasons

- `.pretty` now `.asTable (opts?: { maxColumnWidths:? { callee, file, sourceLine } })`

- It is no longer `extends Array` due to non-working user subclassing with Babel ES5
    - Use `.items` to access original array

- No static methods and properties (except public `.locationsEqual` helper)
    - `.isThirdParty` is now just an overrideable method
    - `.maxColumnWidths` is an overrideable method + a configuration option passed to `asTable ()` method


test.js
"use strict";

/*  ------------------------------------------------------------------------ */

const nodeVersion = Math.floor (Number (process.version.match (/^v(\d+\.\d+)/)[1]))

/*  ------------------------------------------------------------------------ */
                    
require ('chai').should ()

/*  ------------------------------------------------------------------------ */

describe ('impl/partition', () => {

    const partition = require ('./impl/partition')
    const spans     = partition ([ 'a', 'b', 'c', undefined, undefined, 42], x => typeof x)

    spans.should.deep.equal ([ { label: 'string',    items: ['a', 'b', 'c'] },
                               { label: 'undefined', items: [undefined, undefined] },
                               { label: 'number',    items: [42] } ])
})

/*  ------------------------------------------------------------------------ */

describe ('StackTracey', () => {

    const path = require ('path')
    const StackTracey = require ('./stacktracey')
    
    StackTracey.resetCache ()

    const shouldBeVisibleInStackTrace = () => new StackTracey () // @hide

    it ('works', () => {

        const stack = shouldBeVisibleInStackTrace ()

        stack.should.be.an.instanceof (StackTracey)

        stack.items[0].should.deep.equal ({
            beforeParse: 'at shouldBeVisibleInStackTrace (' + path.join (process.cwd (), 'test.js') + ':32:47)',
            callee: 'shouldBeVisibleInStackTrace',
            index: false,
            native: false,
            file: path.join (process.cwd (), 'test.js').replace (/\\/g, '/'),
            line: 32,
            column: 47,
            calleeShort: 'shouldBeVisibleInStackTrace',
            fileName: 'test.js',
            fileRelative: 'test.js',
            fileShort: 'test.js',
            thirdParty: false
        })
    })

    it ('allows to read sources', () => {

        const stack = shouldBeVisibleInStackTrace ().withSources () // @hide

        stack.should.be.an.instanceof (StackTracey)
        stack.items[0].beforeParse.should.not.be.undefined // should preserve previous fields
        stack.items[0].sourceLine.should.equal ('    const shouldBeVisibleInStackTrace = () => new StackTracey () ')
        stack.items[0].hide.should.equal (true) // reads // @hide marker
        stack.items[1].hide.should.equal (true) // reads // @hide marker

        const cleanStack = stack.clean ()

        cleanStack.should.be.an.instanceof (StackTracey)

        StackTracey.locationsEqual (cleanStack.items[0], stack.items[0]).should.equal (true)  // should not clean top element
        StackTracey.locationsEqual (cleanStack.items[1], stack.items[1]).should.equal (false) // should clean second element (due to // @hide)
    })
    

    it ('allows to read sources (async)', () => {

        return shouldBeVisibleInStackTrace ().withSourcesAsync ().then (stack => { // @hide

            stack.should.be.an.instanceof (StackTracey)
            stack.items[0].beforeParse.should.not.be.undefined // should preserve previous fields
            stack.items[0].sourceLine.should.equal ('    const shouldBeVisibleInStackTrace = () => new StackTracey () ')
            stack.items[0].hide.should.equal (true) // reads // @hide marker
            stack.items[1].hide.should.equal (true) // reads // @hide marker

            return stack.cleanAsync ().then (cleanStack => {

                cleanStack.should.be.an.instanceof (StackTracey)

                StackTracey.locationsEqual (cleanStack.items[0], stack.items[0]).should.equal (true)  // should not clean top element
                StackTracey.locationsEqual (cleanStack.items[1], stack.items[1]).should.equal (false) // should clean second element (due to // @hide)
            })
        })
    })

    it ('allows creation from array + groups duplicate lines', () => {

        const stack = new StackTracey ([
            { file: 'yo.js',  line: 11, callee: 'a.funkktion',   calleeShort: 'a' },
            { file: 'yo.js',  line: 10, callee: 'foobar.boobar', calleeShort: 'foobar' },
            { file: 'yo.js',  line: 10, callee: 'foobar.boobar', calleeShort: 'foobar' },
            { file: 'lol.js', line: 10, callee: '',              calleeShort: '' },
        ])

        const clean = stack.clean ().map (x => Object.assign ({
                                                    file: x.file,
                                                    line: x.line,
                                                    callee: x.callee,
                                                    calleeShort: x.calleeShort }))

        clean.should.be.an.instanceof (StackTracey)

        clean.items.should.deep.equal ([ // .should does not recognize StackTracey as normal array...

            { file: 'yo.js',  line: 11, callee: 'a.funkktion',   calleeShort: 'a' },
            { file: 'yo.js',  line: 10, callee: 'foobar.boobar → foobar.boobar', calleeShort: 'foobar → foobar' },
            { file: 'lol.js', line: 10, callee: '',              calleeShort: '' },
        ])
    })

    it ('handles inaccessible files', () => {

        const stack = shouldBeVisibleInStackTrace ()
              stack.items[0].file = '^___^'
              stack.withSources ().items[0].sourceLine.should.equal ('')
              stack.withSourceAt (0).error.should.be.an.instanceof (Error)
    })

    it ('exposes some Array methods', () => {

        const stack = shouldBeVisibleInStackTrace ()
        const sliced = stack.slice (1)
        const deltaLength = (stack.items.length - sliced.items.length)

        deltaLength.should.equal (1)
        sliced.should.be.an.instanceof (StackTracey)

        sliced.filter (x => true).should.be.an.instanceof (StackTracey)
    })

    it ('works with sourcemaps', () => {

        const mkay = require ('./test_files/mkay.uglified')

        try {
            mkay ()
        }
        catch (e) {

            e.message.should.equal ('mkay')

            const top = new StackTracey (e).withSourceAt (0)

            top.line        .should.equal (4)
            top.sourceLine  .should.equal ('\t\t\t\t\tthrow new Error (\'mkay\') }')

            top.file        .should.equal (path.resolve ('./test_files/mkay.js').replace (/\\/g, '/'))
            top.fileShort   .should.equal ('test_files/mkay.js')
            top.fileName    .should.equal ('mkay.js')
        }
    })

    it ('pretty printing works', function prettyTest () {

        const pretty = new StackTracey ().clean ().asTable ()

        console.log ('')
        console.log (pretty, '\n')

        //const spaces = nodeVersion > 8 ? '        ' : '                      ';

        console.log (pretty.split ('\n')[0].match (/at prettyTest\s+test.js\:144\s+const pretty = new StackTracey \(\)\.clean\.pretty/))
    })

    it ('trims too long columns in the pretty printed output', () => {

        const stack = new StackTracey ([
            { fileShort: 'dasdasdasdadadadasdasdasdasdasddasdsadadasdassdasdaddadasdas.js', line: 11, calleeShort: 'dadasdasdasdasdasdasdasdasdasdasdasdasd' },
        ])

        stack.withSources ().asTable ().split ('\n')[0].should.equal ('at dadasdasdasdasdasdasdasdasdas…  …asdadadadasdasdasdasdasddasdsadadasdassdasdaddadasdas.js:11  ')
    })
    
    it ('exposes Array methods', () => {

        const stack = new StackTracey ([
            { file: 'foo' },
            { file: 'bar' }
        ])

        const mapped = stack.map ((x, i) => Object.assign (x, { i }))

        mapped.items.should.deep.equal ([ { file: 'foo', i: 0 }, { file: 'bar', i: 1 } ])
        mapped.should.be.an.instanceof (StackTracey)

        const filtered = stack.filter (x => x.file === 'bar')

        filtered.items.length.should.equal (1)
        filtered.items[0].should.deep.equal ({ file: 'bar', i: 1 })
    })

    it ('computes relative path correctly', () => {
        
        StackTracey.prototype.relativePath ('webpack:///~/jquery/dist/jquery.js')
                             .should.equal (           '~/jquery/dist/jquery.js')

        StackTracey.prototype.relativePath ('webpack:/webpack/bootstrap')
                             .should.equal (         'webpack/bootstrap')
    })

    it ('computes short path correctly', () => {

        StackTracey.prototype.shortenPath  ('webpack/bootstrap/jquery/dist/jquery.js')
                             .should.equal ('jquery/dist/jquery.js')

        StackTracey.prototype.shortenPath  ('node_modules/jquery/dist/jquery.js')
                             .should.equal ('jquery/dist/jquery.js')
    })

    if (nodeVersion >= 5) {

        it ('recognizes SyntaxErrors', () => {

            try { require ('./test_files/syntax_error.js') }
            catch (e) {

                const stack = new StackTracey (e).clean ()

                console.log ('')
                console.log (stack.asTable (), '\n')

                stack.items[0].syntaxError.should.equal (true)
                stack.items[0].column.should.equal (5)
                
                const spaces  = nodeVersion > 8 ? '    ' : '                  '
                const spaces2 = nodeVersion > 8 ? (nodeVersion > 11 ? '         ' : '        ') : '  '

                stack.asTable ().split ('\n')[0].should.equal ('at (syntax error)' + spaces + 'test_files/syntax_error.js:2' + spaces2 + 'foo->bar ()                                     ')
            }
        })
    }

    it ('allows to override isThirdParty()', () => {

        class MyStackTracey extends StackTracey {
            isThirdParty (path) {
                return super.isThirdParty (path) || (path === 'test.js')
            }
        }

        new MyStackTracey ().items[0].thirdParty.should.equal (true)

        class MyStackTracey2 extends MyStackTracey {
            isThirdParty (path) {
                return super.isThirdParty (path) && (path !== 'test.js')
            }
        }
        
        new MyStackTracey2 ().items[0].thirdParty.should.equal (false)
    })

    it ('.withSourceAt', () => {

        const line = new StackTracey ().withSourceAt (0).sourceLine.trim ()
        line.should.equal ('const line = new StackTracey ().withSourceAt (0).sourceLine.trim ()')
    })

    it ('.at', () => {
        
        new StackTracey ().at (0).file.includes ('stacktracey/test.js').should.equal (true)
    })

    it ('detects Array methods as native', () => {

        const arr = [1,2,3]
        const stack = arr.reduce (() => new StackTracey ())

        stack.items[1].native.should.equal (true)
    })

    it ('works on Windows', () => {

        const dir = process.cwd ()

        const windowsStack =
                [
                'Error',
                '    at Context.it (' + dir + '\\test.js:38:22)',
                '    at callFn (' + dir + '\\node_modules\\mocha\\lib\\runnable.js:354:21)',
                '    at runCallback (timers.js:800:20)'
                ].join ('\n')

        const stack = new StackTracey (windowsStack)
        const pretty = stack.withSources ().asTable ()
        const lines = pretty.split ('\n')

        console.log ('')
        console.log (pretty, '\n')

        lines[0].should.equal ('at it           test.js:38                 stack.should.be.an.instanceof (StackTracey)')
        lines[1].indexOf      ('at callFn       mocha/lib/runnable.js:354').should.equal (0)
    })

})





stacktracey.d.ts
declare interface SourceFile {

    path: string
    text: string
    lines: string[]
    error?: Error
}

declare interface Location {

    file:    string
    line?:   number
    column?: number
}

declare interface Entry extends Location {

    beforeParse: string
    callee:      string
    index:       boolean
    native:      boolean

    calleeShort:  string
    fileRelative: string
    fileShort:    string
    fileName:     string
    thirdParty:   boolean

    hide?:       boolean
    sourceLine?: string
    sourceFile?: SourceFile
    error?:      Error
}

declare interface MaxColumnWidths {
    callee:     number
    file:       number
    sourceLine: number
}

declare class StackTracey {

    constructor (input?: Error|string|Entry[], offset?: number)

    items: Entry[]

    extractEntryMetadata (e: Entry)

    shortenPath (relativePath: string): string
    relativePath (fullPath: string): string
    isThirdParty (relativePath: string): boolean

    rawParse (str: string): Entry[]

    withSourceAt (i: number): Entry
    withSourceAsyncAt (i: number): Promise<Entry>

    withSource (entry: Entry): Entry
    withSourceAsync (entry: Entry): Promise<Entry>

    withSources ():        StackTracey
    withSourcesAsync ():   Promise<StackTracey>
    mergeRepeatedLines (): StackTracey

    clean ():              StackTracey
    cleanAsync ():         Promise<StackTracey>

    isClean (entry: Entry, index: number): boolean

    map    (f: (x: Entry, i: number, arr: Entry[]) => Entry): StackTracey
    filter (f: (x: Entry, i: number, arr: Entry[]) => Entry): StackTracey
    slice  (from?: number, to?: number):                      StackTracey
    concat (...args: Entry[]):                                StackTracey

    at (i: number): Entry

    asTable (opts?: { maxColumnWidths?: MaxColumnWidths }): string

    maxColumnWidths (): MaxColumnWidths

    static resetCache (): void
    static locationsEqual (a: Location, b: Location): boolean
}

export = StackTracey


stacktracey.js
"use strict";

/*  ------------------------------------------------------------------------ */

const O              = Object,
      isBrowser      = (typeof window !== 'undefined') && (window.window === window) && window.navigator,
      lastOf         = x => x[x.length - 1],
      getSource      = require ('get-source'),
      partition      = require ('./impl/partition'),
      asTable        = require ('as-table'),
      nixSlashes     = x => x.replace (/\\/g, '/'),
      pathRoot       = isBrowser ? window.location.href : (nixSlashes (process.cwd ()) + '/'),
      pathToRelative = isBrowser ? ((root, full) => full.replace (root, '')) : module.require ('path').relative

/*  ------------------------------------------------------------------------ */

class StackTracey {

    constructor (input, offset) {
        
        const originalInput          = input
            , isParseableSyntaxError = input && (input instanceof SyntaxError && !isBrowser)
                
    /*  new StackTracey ()            */

        if (!input) {
             input = new Error ()
             offset = (offset === undefined) ? 1 : offset
        }

    /*  new StackTracey (Error)      */

        if (input instanceof Error) {
            input = input.stack || ''
        }

    /*  new StackTracey (string)     */

        if (typeof input === 'string') {
            input = this.rawParse (input).slice (offset).map (x => this.extractEntryMetadata (x))
        }

    /*  new StackTracey (array)      */

        if (Array.isArray (input)) {

            if (isParseableSyntaxError) {
                
                const rawLines = module.require ('util').inspect (originalInput).split ('\n')
                    , fileLine = rawLines[0].split (':')
                    , line = fileLine.pop ()
                    , file = fileLine.join (':')

                if (file) {
                    input.unshift ({
                        file: nixSlashes (file),
                        line: line,
                        column: (rawLines[2] || '').indexOf ('^') + 1,
                        sourceLine: rawLines[1],
                        callee: '(syntax error)',
                        syntaxError: true
                    })
                }
            }

            this.items = input

        } else {
            this.items = []
        }
    }

    extractEntryMetadata (e) {

        const fileRelative = this.relativePath (e.file || '')

        return O.assign (e, {

            calleeShort:  e.calleeShort || lastOf ((e.callee || '').split ('.')),
            fileRelative: fileRelative,
            fileShort:    this.shortenPath (fileRelative),
            fileName:     lastOf ((e.file || '').split ('/')),
            thirdParty:   this.isThirdParty (fileRelative) && !e.index
        })
    }

    shortenPath (relativePath) {
        return relativePath.replace (/^node_modules\//, '')
                           .replace (/^webpack\/bootstrap\//, '')
    }

    relativePath (fullPath) {
        return nixSlashes (pathToRelative (pathRoot, fullPath)).replace (/^.*\:\/\/?\/?/, '')
    }

    isThirdParty (relativePath) {
        return (relativePath[0] === '~')                          || // webpack-specific heuristic
               (relativePath[0] === '/')                          || // external source
               (relativePath.indexOf ('node_modules')      === 0) ||
               (relativePath.indexOf ('webpack/bootstrap') === 0)
    }

    rawParse (str) {

        const lines = (str || '').split ('\n')

        const entries = lines.map (line => {

            line = line.trim ()

            let callee, fileLineColumn = [], native, planA, planB

            if ((planA = line.match (/at (.+) \((.+)\)/)) ||
                ((line.slice (0, 3) !== 'at ') && (planA = line.match (/(.*)@(.*)/)))) {

                callee         =  planA[1]
                native         = (planA[2] === 'native')
                fileLineColumn = (planA[2].match (/(.*):(.+):(.+)/) || []).slice (1)

            } else if ((planB = line.match (/^(at\s+)*(.+):([0-9]+):([0-9]+)/) )) {
                fileLineColumn = (planB).slice (2)

            } else {
                return undefined
            }

        /*  Detect things like Array.reduce
            TODO: detect more built-in types            */
            
            if (callee && !fileLineColumn[0]) {
                const type = callee.split ('.')[0]
                if (type === 'Array') {
                    native = true
                }
            }

            return {
                beforeParse: line,
                callee:      callee || '',
                index:       isBrowser && (fileLineColumn[0] === window.location.href),
                native:      native || false,
                file:        nixSlashes (fileLineColumn[0] || ''),
                line:        parseInt (fileLineColumn[1] || '', 10) || undefined,
                column:      parseInt (fileLineColumn[2] || '', 10) || undefined
            }
        })

        return entries.filter (x => (x !== undefined))
    }

    withSourceAt (i) {
        return this.items[i] && this.withSource (this.items[i])
    }

    withSourceAsyncAt (i) {
        return this.items[i] && this.withSourceAsync (this.items[i])
    }

    withSource (loc) {
        
        if (this.shouldSkipResolving (loc)) {
            return loc
            
        } else {

            let resolved = getSource (loc.file || '').resolve (loc)

            if (!resolved.sourceFile) {
                return loc
            }

            return this.withSourceResolved (loc, resolved)
        }
    }

    withSourceAsync (loc) {

        if (this.shouldSkipResolving (loc)) {
            return Promise.resolve (loc)
            
        } else {
            return getSource.async (loc.file || '')
                        .then (x => x.resolve (loc))
                        .then (resolved => this.withSourceResolved (loc, resolved))
                        .catch (e => this.withSourceResolved (loc, { error: e, sourceLine: '' }))
        }
    }

    shouldSkipResolving (loc) {
        return loc.sourceFile || loc.error || (loc.file && loc.file.indexOf ('<') >= 0) // skip things like <anonymous> and stuff that was already fetched
    }

    withSourceResolved (loc, resolved) {

        if (resolved.sourceFile && !resolved.sourceFile.error) {
            resolved.file = nixSlashes (resolved.sourceFile.path)
            resolved = this.extractEntryMetadata (resolved)
        }

        if (resolved.sourceLine.includes ('// @hide')) {
            resolved.sourceLine = resolved.sourceLine.replace  ('// @hide', '')
            resolved.hide       = true
        }

        if (resolved.sourceLine.includes ('__webpack_require__') || // webpack-specific heuristics
            resolved.sourceLine.includes ('/******/ ({')) {
            resolved.thirdParty = true
        }

        return O.assign ({ sourceLine: '' }, loc, resolved)
    }

    withSources () {
        return this.map (x => this.withSource (x))
    }

    withSourcesAsync () {
        return Promise.all (this.items.map (x => this.withSourceAsync (x)))
                      .then (items => new StackTracey (items))
    }

    mergeRepeatedLines () {
        return new StackTracey (
            partition (this.items, e => e.file + e.line).map (
                group => {
                    return group.items.slice (1).reduce ((memo, entry) => {
                        memo.callee      = (memo.callee      || '<anonymous>') + ' → ' + (entry.callee      || '<anonymous>')
                        memo.calleeShort = (memo.calleeShort || '<anonymous>') + ' → ' + (entry.calleeShort || '<anonymous>')
                        return memo
                    }, O.assign ({}, group.items[0]))
                }
            )
        )
    }

    clean () {
        const s = this.withSources ().mergeRepeatedLines ()
        return s.filter (s.isClean.bind (s))
    }

    cleanAsync () {
        return this.withSourcesAsync ().then (s => {
            s = s.mergeRepeatedLines ()
            return s.filter (s.isClean.bind (s))
        })
    }

    isClean (entry, index) {
        return (index === 0) || !(entry.thirdParty || entry.hide || entry.native)
    }

    at (i) {
        return O.assign ({

            beforeParse: '',
            callee:      '<???>',
            index:       false,
            native:      false,
            file:        '<???>',
            line:        0,
            column:      0

        }, this.items[i])
    }

    asTable (opts) {

        const maxColumnWidths = (opts && opts.maxColumnWidths) || this.maxColumnWidths ()

        const trimEnd   = (s, n) => s && ((s.length > n) ? (s.slice (0, n-1) + '…') : s)   
        const trimStart = (s, n) => s && ((s.length > n) ? ('…' + s.slice (-(n-1))) : s)

        const trimmed = this.map (
            e => [
                ('at ' + trimEnd (e.calleeShort,                                maxColumnWidths.callee)),
                trimStart ((e.fileShort && (e.fileShort + ':' + e.line)) || '', maxColumnWidths.file),
                trimEnd (((e.sourceLine || '').trim () || ''),                  maxColumnWidths.sourceLine)
            ]
        )

        return asTable (trimmed.items)
    }

    maxColumnWidths () {
        return {
            callee:     30,
            file:       60,
            sourceLine: 80
        }
    }

    static resetCache () {

        getSource.resetCache ()
        getSource.async.resetCache ()
    }

    static locationsEqual (a, b) {

        return (a.file   === b.file) &&
               (a.line   === b.line) &&
               (a.column === b.column)
    }
}

/*  Array methods
    ------------------------------------------------------------------------ */

;['map', 'filter', 'slice', 'concat'].forEach (method => {

    StackTracey.prototype[method] = function (/*...args */) { // no support for ...args in Node v4 :(
        return new StackTracey (this.items[method].apply (this.items, arguments))
    }
})

/*  ------------------------------------------------------------------------ */

module.exports = StackTracey

